src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/EmbeddedLanguages/VirtualChars/AbstractVirtualCharService.cs

```diff
@@ -159,25 +159,16 @@ private void CheckInvariants(SyntaxToken token, VirtualCharSequence result)
                     result.Add(VirtualChar.Create(new Rune(tokenText[index]), span));
                     index += result.Last().Span.Length;
                 }
-                else if (Rune.TryCreate(tokenText[index], out var rune))
-                {
-                    // First, see if this was a single char that can become a rune (the common case).
-                    result.Add(VirtualChar.Create(rune, new TextSpan(offset + index, 1)));
-                    index += 1;
-                }
-                else if (index + 1 < tokenText.Length &&
-                         Rune.TryCreate(tokenText[index], tokenText[index + 1], out rune))
-                {
-                    // Otherwise, see if we have a surrogate pair (less common, but possible).
-                    result.Add(VirtualChar.Create(rune, new TextSpan(offset + index, 2)));
-                    index += 2;
-                }
                 else
                 {
-                    // Something that couldn't be encoded as runes.
-                    Debug.Assert(char.IsSurrogate(tokenText[index]));
-                    result.Add(VirtualChar.Create(tokenText[index], new TextSpan(offset + index, 1)));
-                    index += 1;
+                    var virtualChar = VirtualChar.CreateNextInString(
+                        tokenText,
+                        index,
+                        (start, length) => new TextSpan(offset + index, length),
```

Comment:
not happy about a lambda here.

Reply:
this is called for every char in the string.

Reply:
can you not just pass in the offset to CreateNextInString ?

Reply:
Sure, I'll change to take in an offset